File: discord_bot/routes/lounge.py
# routes/lounge.py
from quart import Blueprint, current_app, render_template, request, session, redirect, url_for, jsonify
from services.lounge_service import LoungeService
from services.tournament_service import TitleService
from services.bridge_client import BridgeUnavailableError
from routes.tournament import _ensure_discord_role, _assign_title_role

lounge_bp = Blueprint("lounge", __name__, url_prefix="/lounge")


def _current_user():
    return session.get("discord_user")


def _require_login():
    user = _current_user()
    if not user:
        return redirect(url_for("login"))
    return None


def _is_host(user, session_data) -> bool:
    return str(user["id"]) == str(session_data.get("host_id", ""))


# ============================================================
# ラウンジ画面
# ============================================================

@lounge_bp.route("/")
async def index():
    redir = _require_login()
    if redir:
        return redir
    user = _current_user()
    try:
        active_sessions = await LoungeService.list_active_sessions()
    except Exception:
        active_sessions = []
    return await render_template("lounge.html", user=user, view="list", active_sessions=active_sessions)


@lounge_bp.route("/sessions/<int:session_id>")
async def session_view(session_id: int):
    redir = _require_login()
    if redir:
        return redir
    user = _current_user()
    try:
        session_data = await LoungeService.get_session(session_id)
        if not session_data:
            return redirect(url_for("lounge.index"))
        final_scores = await LoungeService.get_final_scores(session_id)
        members = await LoungeService.list_members(session_id)
        is_host = _is_host(user, session_data)
        return await render_template(
            "lounge.html",
            user=user,
            lounge_session=session_data,
            final_scores=final_scores,
            members=members,
            view="session",
            is_host=is_host,
        )
    except BridgeUnavailableError:
        return await render_template("maintenance.html"), 503


# ============================================================
# API（Ajax用）
# ============================================================

@lounge_bp.route("/api/sessions", methods=["POST"])
async def api_create_session():
    user = _current_user()
    if not user:
        return jsonify({"status": "error"}), 401
    data = await request.get_json()
    session_id = await LoungeService.create_session(
        room_id=data.get("room_id", ""),
        host_id=int(user["id"]),
        mode=data.get("mode", "ffa"),
        total_races=int(data.get("total_races", 12)),
    )
    return jsonify({"status": "ok" if session_id else "error", "session_id": session_id})


@lounge_bp.route("/api/sessions/<int:session_id>/join", methods=["POST"])
async def api_join_session(session_id: int):
    user = _current_user()
    if not user:
        return jsonify({"status": "error"}), 401
    ok = await LoungeService.add_member(session_id, int(user["id"]))
    return jsonify({"status": "ok" if ok else "error"})


@lounge_bp.route("/api/sessions/<int:session_id>/final-scores/report", methods=["POST"])
async def api_report_final_score(session_id: int):
    user = _current_user()
    if not user:
        return jsonify({"status": "error"}), 401
    data = await request.get_json()
    final_rank = data.get("final_rank")
    if final_rank is None or not (1 <= int(final_rank) <= 24):
        return jsonify({"status": "error", "message": "final_rank は 1〜24 で指定してください"}), 400
    ok = await LoungeService.report_final_score(session_id, int(user["id"]), int(final_rank))
    return jsonify({"status": "ok" if ok else "error"})


@lounge_bp.route("/api/sessions/<int:session_id>/final-scores")
async def api_get_final_scores(session_id: int):
    if not _current_user():
        return jsonify([]), 401
    scores = await LoungeService.get_final_scores(session_id)
    return jsonify(scores)


@lounge_bp.route("/api/sessions/<int:session_id>/exclude", methods=["POST"])
async def api_exclude_player(session_id: int):
    user = _current_user()
    if not user:
        return jsonify({"status": "error"}), 401
    session_data = await LoungeService.get_session(session_id)
    if not session_data or not _is_host(user, session_data):
        return jsonify({"status": "error", "message": "ホストのみ操作できます"}), 403
    data = await request.get_json()
    excluded = await LoungeService.exclude_player(session_id, int(data.get("user_id")))
    if excluded is None:
        return jsonify({"status": "error"}), 500
    return jsonify({"status": "ok", "excluded": excluded})


async def _do_finish_session(session_id: int):
    """セッション終了フロー（MMR計算 + 称号付与 + Discordロール同期）。"""
    res = await LoungeService.finish_session(session_id)
    if not res or res.get("status") != "ok":
        return False

    # Bridge が MMR 計算済みの結果を返す
    results = res.get("results", [])
    all_titles = await TitleService.list_all()

    for entry in results:
        uid = entry.get("user_id")
        new_mmr = entry.get("new_mmr", 0)
        if not uid:
            continue
        newly_granted_ids = await TitleService.grant_rank(uid, new_mmr)
        for title_id in newly_granted_ids:
            active = await TitleService.get_active(uid)
            granted_title = next((t for t in all_titles if t["id"] == title_id), None)
            if granted_title:
                await _ensure_discord_role(granted_title)
                if active is None:
                    await TitleService.set_active(uid, title_id)
                await _assign_title_role(str(uid), granted_title)
    return True


@lounge_bp.route("/api/sessions/<int:session_id>/finish", methods=["POST"])
async def api_finish_session(session_id: int):
    user = _current_user()
    if not user:
        return jsonify({"status": "error"}), 401
    session_data = await LoungeService.get_session(session_id)
    if not session_data or not _is_host(user, session_data):
        return jsonify({"status": "error", "message": "ホストのみ操作できます"}), 403
    ok = await _do_finish_session(session_id)
    return jsonify({"status": "ok" if ok else "error"})


@lounge_bp.route("/api/sessions/<int:session_id>/my-result")
async def api_my_result(session_id: int):
    """セッション終了後の個人結果（最終順位・MMR増加・現在MMR・称号）を返す。"""
    user = _current_user()
    if not user:
        return jsonify({}), 401
    uid_str = str(user["id"])

    final_scores = await LoungeService.get_final_scores(session_id)
    my_entry = next((s for s in final_scores if str(s.get("user_id")) == uid_str), None)

    final_rank = my_entry.get("final_rank") if my_entry else None
    mmr_delta = my_entry.get("mmr_delta", 0) if my_entry else 0
    total_submitted = sum(1 for s in final_scores if s.get("submitted"))

    player = await LoungeService.get_player(int(user["id"]))
    current_mmr = player.get("mmr", 1000) if player else 1000

    all_titles = await TitleService.list_all()
    player_titles = await TitleService.get_player_titles(int(user["id"]))
    earned_ids = {t["id"] for t in player_titles if t.get("earned")}
    active_title_name = None
    for t in sorted(all_titles, key=lambda x: x.get("unlock_threshold") or 0, reverse=True):
        if t["id"] in earned_ids and t.get("unlock_type") == "lounge_rank":
            active_title_name = t["name"]
            break

    return jsonify({
        "final_rank":      final_rank,
        "total_players":   total_submitted,
        "mmr_delta":       mmr_delta,
        "mmr":             current_mmr,
        "rank_name":       active_title_name or "—",
        "is_winner":       final_rank == 1,
    })


@lounge_bp.route("/api/sessions/<int:session_id>/standings")
async def api_standings(session_id: int):
    if not _current_user():
        return jsonify([])
    standings = await LoungeService.get_standings(session_id)
    return jsonify(standings)


@lounge_bp.route("/api/me")
async def api_me():
    """ログインユーザーの MMR とランク称号を返す。"""
    user = _current_user()
    if not user:
        return jsonify({}), 401
    uid = int(user["id"])
    player = await LoungeService.get_player(uid)
    titles = await TitleService.get_player_titles(uid)
    rank_title = None
    for t in sorted(titles, key=lambda x: x.get("unlock_threshold") or 0, reverse=True):
        if t.get("unlock_type") == "lounge_rank" and t.get("earned"):
            rank_title = t.get("name")
            break
    mmr = player.get("mmr", 1000) if player else 1000
    return jsonify({"mmr": mmr, "rank_name": rank_title or "—"})


@lounge_bp.route("/api/sessions/<int:session_id>/teams", methods=["GET", "POST"])
async def api_teams(session_id: int):
    user = _current_user()
    if not user:
        return jsonify({"status": "error"}), 401
    if request.method == "POST":
        data = await request.get_json()
        team_id = await LoungeService.create_team(
            session_id, data.get("tag", ""), data.get("member_ids", [])
        )
        return jsonify({"status": "ok" if team_id else "error", "team_id": team_id})
    teams = await LoungeService.list_teams(session_id)
    return jsonify(teams)

File: database_bridge/src/api/handlers/lounge.rs
// api/handlers/lounge.rs
// Phase 3: セッション最終順位申告方式。
use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::Json;
use serde::Deserialize;
use serde_json::{json, Value};
use sqlx::MySqlPool;

use crate::api::AppState;
use crate::db::{lounge_repo, models::BridgeError};

fn map_err(err: BridgeError) -> (StatusCode, Json<Value>) {
    match err {
        BridgeError::NotFound(msg) => (StatusCode::NOT_FOUND, Json(json!({"status":"error","message":msg}))),
        BridgeError::PermissionDenied => (StatusCode::FORBIDDEN, Json(json!({"status":"error","message":"permission denied"}))),
        BridgeError::Sqlx(e) => {
            tracing::error!("DB error: {}", e);
            (StatusCode::INTERNAL_SERVER_ERROR, Json(json!({"status":"error","message":e.to_string()})))
        },
        BridgeError::Json(e) => (StatusCode::BAD_REQUEST, Json(json!({"status":"error","message":format!("invalid json: {}", e)}))),
    }
}

// ============================================================
// POST /lounge/sessions
// ============================================================
#[derive(Deserialize)]
pub struct CreateSessionRequest {
    pub room_id: String,
    pub mode: Option<String>,
    pub total_races: Option<i8>,
    pub host_id: i64,
}

pub async fn create_session(
    State(pool): State<MySqlPool>,
    Json(payload): Json<CreateSessionRequest>,
) -> (StatusCode, Json<Value>) {
    let mode = payload.mode.as_deref().unwrap_or("ffa");
    let total_races = payload.total_races.unwrap_or(12);
    match lounge_repo::create_session(&pool, &payload.room_id, mode, total_races, payload.host_id).await {
        Ok(id) => (StatusCode::OK, Json(json!({"status":"ok","session_id":id}))),
        Err(e) => map_err(e),
    }
}

// ============================================================
// GET /lounge/sessions
// ============================================================
pub async fn list_sessions(State(pool): State<MySqlPool>) -> (StatusCode, Json<Value>) {
    match lounge_repo::list_active_sessions(&pool).await {
        Ok(sessions) => (StatusCode::OK, Json(json!(sessions))),
        Err(e) => map_err(e),
    }
}

// ============================================================
// GET /lounge/sessions/{id}
// ============================================================
pub async fn get_session(
    State(pool): State<MySqlPool>,
    Path(session_id): Path<i64>,
) -> (StatusCode, Json<Value>) {
    match lounge_repo::get_session(&pool, session_id).await {
        Ok(session) => (StatusCode::OK, Json(json!(session))),
        Err(e) => map_err(e),
    }
}

// ============================================================
// POST /lounge/sessions/{id}/members
// ============================================================
#[derive(Deserialize)]
pub struct AddMemberRequest {
    pub user_id: i64,
}

pub async fn add_member(
    State(pool): State<MySqlPool>,
    Path(session_id): Path<i64>,
    Json(payload): Json<AddMemberRequest>,
) -> (StatusCode, Json<Value>) {
    match lounge_repo::add_session_member(&pool, session_id, payload.user_id).await {
        Ok(_) => (StatusCode::OK, Json(json!({"status":"ok"}))),
        Err(e) => map_err(e),
    }
}

// ============================================================
// GET /lounge/sessions/{id}/members
// ============================================================
pub async fn list_members(
    State(pool): State<MySqlPool>,
    Path(session_id): Path<i64>,
) -> (StatusCode, Json<Value>) {
    match lounge_repo::list_session_members(&pool, session_id).await {
        Ok(members) => (StatusCode::OK, Json(json!(members))),
        Err(e) => map_err(e),
    }
}

// ============================================================
// POST /lounge/sessions/{id}/exclude
// ============================================================
#[derive(Deserialize)]
pub struct ExcludePlayerRequest {
    pub user_id: i64,
}

pub async fn exclude_player(
    State(state): State<AppState>,
    Path(session_id): Path<i64>,
    Json(payload): Json<ExcludePlayerRequest>,
) -> (StatusCode, Json<Value>) {
    match lounge_repo::toggle_exclude_player(&state.pool, session_id, payload.user_id).await {
        Ok(excluded) => {
            let _ = state.tx.send(json!({
                "type":       "lounge.member_excluded",
                "session_id": session_id,
                "user_id":    payload.user_id.to_string(),
                "excluded":   excluded,
            }).to_string());
            (StatusCode::OK, Json(json!({"status":"ok","excluded":excluded})))
        },
        Err(e) => map_err(e),
    }
}

// ============================================================
// POST /lounge/sessions/{id}/final-scores/report
// ============================================================
#[derive(Deserialize)]
pub struct ReportFinalScoreRequest {
    pub user_id: i64,
    pub final_rank: i8,
}

pub async fn report_final_score(
    State(state): State<AppState>,
    Path(session_id): Path<i64>,
    Json(payload): Json<ReportFinalScoreRequest>,
) -> (StatusCode, Json<Value>) {
    match lounge_repo::report_final_score(&state.pool, session_id, payload.user_id, payload.final_rank).await {
        Ok(_) => {
            let _ = state.tx.send(json!({
                "type":       "lounge.final_score_reported",
                "session_id": session_id,
                "user_id":    payload.user_id.to_string(),
                "final_rank": payload.final_rank,
            }).to_string());
            (StatusCode::OK, Json(json!({"status":"ok"})))
        },
        Err(e) => map_err(e),
    }
}

// ============================================================
// GET /lounge/sessions/{id}/final-scores
// ============================================================
pub async fn get_final_scores(
    State(pool): State<MySqlPool>,
    Path(session_id): Path<i64>,
) -> (StatusCode, Json<Value>) {
    match lounge_repo::get_final_scores(&pool, session_id).await {
        Ok(scores) => (StatusCode::OK, Json(json!(scores))),
        Err(e) => map_err(e),
    }
}

// ============================================================
// GET /lounge/sessions/{id}/standings
// ============================================================
pub async fn get_standings(
    State(pool): State<MySqlPool>,
    Path(session_id): Path<i64>,
) -> (StatusCode, Json<Value>) {
    match lounge_repo::get_session_standings(&pool, session_id).await {
        Ok(standings) => (StatusCode::OK, Json(json!(standings))),
        Err(e) => map_err(e),
    }
}

// ============================================================
// POST /lounge/sessions/{id}/finish
// ============================================================
pub async fn finish_session(
    State(state): State<AppState>,
    Path(session_id): Path<i64>,
) -> (StatusCode, Json<Value>) {
    // MMR 計算・更新（申告済み・非除外プレイヤーのみ）
    let results = match lounge_repo::calc_and_apply_mmr(&state.pool, session_id).await {
        Ok(r) => r,
        Err(e) => return map_err(e),
    };

    // セッションを finished に更新し total_sessions をインクリメント
    if let Err(e) = lounge_repo::finish_session(&state.pool, session_id).await {
        return map_err(e);
    }

    // 称号付与・Discordロール同期はPython側で行う
    let _ = state.tx.send(json!({
        "type":       "lounge.session_finished",
        "session_id": session_id,
        "results":    results,
    }).to_string());

    (StatusCode::OK, Json(json!({"status":"ok","results":results})))
}

// ============================================================
// GET /lounge/players/{user_id}
// ============================================================
pub async fn get_player(
    State(pool): State<MySqlPool>,
    Path(user_id): Path<i64>,
) -> (StatusCode, Json<Value>) {
    match lounge_repo::get_lounge_player(&pool, user_id).await {
        Ok(p) => (StatusCode::OK, Json(json!({
            "user_id":       p.user_id,
            "mmr":           p.mmr,
            "peak_mmr":      p.peak_mmr,
            "total_races":   p.total_races,
            "total_sessions": p.total_sessions,
        }))),
        Err(e) => map_err(e),
    }
}

// ============================================================
// POST /lounge/sessions/{id}/teams
// ============================================================
#[derive(Deserialize)]
pub struct CreateTeamRequest {
    pub tag: String,
    pub member_ids: Vec<i64>,
}

pub async fn create_team(
    State(pool): State<MySqlPool>,
    Path(session_id): Path<i64>,
    Json(payload): Json<CreateTeamRequest>,
) -> (StatusCode, Json<Value>) {
    let team_id = match lounge_repo::create_team(&pool, session_id, &payload.tag).await {
        Ok(id) => id,
        Err(e) => return map_err(e),
    };
    for uid in &payload.member_ids {
        if let Err(e) = lounge_repo::add_team_member(&pool, team_id, *uid).await {
            return map_err(e);
        }
    }
    (StatusCode::OK, Json(json!({"status":"ok","team_id":team_id})))
}

// ============================================================
// GET /lounge/sessions/{id}/teams
// ============================================================
pub async fn list_teams(
    State(pool): State<MySqlPool>,
    Path(session_id): Path<i64>,
) -> (StatusCode, Json<Value>) {
    match lounge_repo::list_teams(&pool, session_id).await {
        Ok(teams) => (StatusCode::OK, Json(json!(teams))),
        Err(e) => map_err(e),
    }
}
